src/main/java/ch/njol/skript/expressions/ExprPermissions.java

```diff
@@ -18,71 +18,135 @@
+
 	static {
-		Skript.registerExpression(ExprPermissions.class, String.class, ExpressionType.PROPERTY, "[(all [[of] the]|the)] permissions (from|of) %players%", "[(all [[of] the]|the)] %players%'[s] permissions");
+		Skript.registerExpression(ExprPermissions.class, String.class, ExpressionType.PROPERTY,
+				"[all [[of] the]|the] permissions (from|of) %entities%",
+				"[all [of] the] %entities%'[s] permissions"
+		);
 	}
-	
-	@SuppressWarnings("null")
-	private Expression<Player> players;
-	
-	@SuppressWarnings({"null", "unchecked"})
+
 	@Override
-	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
-		players = (Expression<Player>) exprs[0];
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		setExpr((Expression<Entity>) exprs[0]);
 		return true;
 	}
 
 	@Override
-	@Nullable
-	protected String[] get(Event e) {
-		final Set<String> permissions = new HashSet<>();
-		for (Player player : players.getArray(e))
-			for (final PermissionAttachmentInfo permission : player.getEffectivePermissions())
-				permissions.add(permission.getPermission());
-		return permissions.toArray(new String[permissions.size()]);
+	protected String[] get(Event event, Entity[] source) {
+		return getExpr().stream(event)
+				.flatMap(permissible -> permissible.getEffectivePermissions().stream())
+				.map(permission -> permission.getPermission())
+				.toArray(String[]::new);
+	}
+
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.ADD || mode == ChangeMode.REMOVE || mode == ChangeMode.DELETE)
+			return CollectionUtils.array(String[].class);
+		return null;
+	}
+
+	@Override
+	public void change(Event event, @Nullable Object[] delta, ChangeMode mode) {
+		List<String> permissions = new ArrayList<>();
+		if (mode != ChangeMode.DELETE) {
+			if (delta == null)
+				return;
+			for (Object permission : delta) {
+				if (permission == null)
+					continue;
+				permissions.add((String) permission);
+			}
+		}
+		if (mode == ChangeMode.DELETE || mode == ChangeMode.REMOVE) {
+			for (Entity entity : getExpr().getArray(event)) {
+				for (PermissionAttachmentInfo info : entity.getEffectivePermissions()) {
+					PermissionAttachment attachment = info.getAttachment();
+					if (attachment == null)
+						continue;
+					for (String permission : attachment.getPermissions().keySet()) {
+						if (mode == ChangeMode.DELETE) {
+							attachment.unsetPermission(permission);
+						} else if (permissions.contains(permission)) {
+							attachment.unsetPermission(permission);
+						}
+					}
+				}
+			}
+			return;
+		}
+		for (Entity entity : getExpr().getAll(event)) {
+			PermissionAttachment attachment = getPermission(entity);
+			for (String permission : permissions)
+				attachment.setPermission(permission, true);
+		}
```

Comment:
This whole implementation is a mess and hard to follow. You should check for the `mode` in a switch statement and neatly put each mode's implementation in its own case.